plugins/ag-charts-generate-chart-thumbnail/src/executors/generate/generator/thumbnailGenerator.ts

```diff
@@ -19,59 +20,111 @@ interface Params {
+    };
+};
+
 export async function generateExample({ name, example, theme, outputPath }: Params) {
     const { entryFileName, files = {} } = example;
 
-    const poolEntry = await borrowFromPool();
-    const [jsdom, mockCtx, chartProxy] = poolEntry;
-
     try {
         const entryFile = files[entryFileName!];
-        let {
-            options,
-            options: { width, height, padding },
-        } = transformPlainEntryFile(entryFile, files['data.js'], theme);
-        width ??= DEFAULT_THUMBNAIL_WIDTH;
-        height ??= DEFAULT_THUMBNAIL_HEIGHT;
-
-        // If there is a custom width/height try to maintain the right aspect ratio.
-        if (width != DEFAULT_THUMBNAIL_WIDTH && height != DEFAULT_THUMBNAIL_HEIGHT) {
-            const defaultAspectRatio = DEFAULT_THUMBNAIL_WIDTH / DEFAULT_THUMBNAIL_HEIGHT;
-            const aspectRatio = width / height;
-            if (aspectRatio >= defaultAspectRatio) {
-                width = DEFAULT_THUMBNAIL_WIDTH;
-                height = Math.round(width / aspectRatio);
-            } else {
-                height = DEFAULT_THUMBNAIL_HEIGHT;
-                width = Math.round(height * aspectRatio);
+        const { optionsById } = transformPlainEntryFile(entryFile, files['data.js'], theme);
+
+        const jsdom = new JSDOM(`<html><head><style></style></head><body>${files['index.html']}</body></html>`);
+        const { window } = jsdom;
+        const { document } = window;
+
+        const outputCanvas = new Canvas(DEFAULT_THUMBNAIL_WIDTH, DEFAULT_THUMBNAIL_HEIGHT);
+        const outputCtx = outputCanvas.getContext('2d');
+
+        const chartContainers = Array.from(document.querySelectorAll<HTMLElement>('[id]'));
+
+        let rows = 1;
+        let columns = 1;
+        chartContainers.forEach((container) => {
+            const gridArea = getGridArea(container);
+            if (gridArea == null) return;
+            rows = Math.max(gridArea.row + 1, rows);
+            columns = Math.max(gridArea.column + 1, columns);
+        });
+
+        const promises = chartContainers.map(async (container) => {
+            const gridArea = getGridArea(container);
+            const row = gridArea?.row ?? 0;
+            const column = gridArea?.column ?? 0;
+
+            const width = (DEFAULT_THUMBNAIL_WIDTH / columns) | 0;
+            const height = (DEFAULT_THUMBNAIL_HEIGHT / rows) | 0;
+            const x0 = (width * column) | 0;
+            const y0 = (width * row) | 0;
+
+            const mockCtx = mockCanvas.setup({
+                width,
+                height,
+                document,
+                window: window as any,
+                mockText: false,
+            });
+
+            const chartOptions = {
+                animation: { enabled: false },
+                document,
+                window,
+                width,
+                height,
+            } as any;
+
+            const chartProxy = AgCharts.create(chartOptions);
```

Comment:
I have some concerns about this approach - every `AgCharts.create()` call is allocating around a dozen OffScreenCanvases (which are then deallocated on `destroy()`), the pool we had previously was minimising the memory churn by re-using the `AgChartInstance`s.

(This was one of the pre-Nx-task-breakout optimisations that made a big difference to the Astro build)